<a href="https://colab.research.google.com/github/averyleiss/NUWIT-website-hack-day-2022/blob/main/Mini_Proj_AL_%2B_TM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Mini Project 2 (22 points total)

Complete the Jupyter Notebook below with your group of two to four students.  You need only submit one Jupyter Notebook to [gradescope](https://www.gradescope.com/courses/347537/assignments/2012798/) for your whole group (be sure to add all group members).  

While you may submit your groups work to gradescope anytime until **April 26 @ 11:59PM** my expectation is that most of the work can be completed in class today.

________________


In this Mini Project, you'll write two homework problems of your choosing.  

# Can future courses re-use these problems? (please?)
We'd love to be able to use these problems in a future semester.  If you'd allow us, please save this jupyter notebook and, in addition to submitting on gradescope, upload [here](https://www.dropbox.com/request/wdM7nfvpxGUvKXdmUqgu).  (This is entirely optional)

We're happy to cite your team, our source for these problems.  If you'd like to be acknowledged, please include your email addresses in the cell below.  You're welcome to leave the cell below blank if you prefer to not have your emails associated with your problem.

Know that we'll very likely edit a problem before sharing with future students.

In [ ]:
# emails to be included

# Data Sources

- [google dataset search](https://datasetsearch.research.google.com/)
- [kaggle](https://kaggle.com/)
- [Global Health Observatory](https://apps.who.int/gho/data/node.home)
- [UCI Machine Learning Repo](https://archive.ics.uci.edu/ml/index.php)
- [data.gov](https://www.data.gov/)
- [FiveThirtyEight](https://data.fivethirtyeight.com/)


# Part 1: Hypothesis Test / Cov&Corr / theory problem

Choose any application of interest and write a multi-part homework problem which practices it.  The best problems:
- have clean connections between the "application story" and the mathematics
- are fun and interesting to think about

Below we list some popular topics for homework problems:

- **hypothesis tests:**
    - t-test
        - is it 1 or 2 sided?
    - chi square test
        - what is the target distribution you're comparing to?
            - some known discrete distribution
            - binomial
            - poisson
            - (whatever the target distribution you're comparing to, be sure its reasonable)
- hypothesis test considerations:
    - can you source real data for this data or must you make up numbers?
    - be sure the problem can be done with pencil / paper
        - limit sample sizes to what can be computed by hand

- **covariance / correlation**
    - identify some interesting question which can be answered from observing correlation / covariance
    - you must source real data for this
    - once you have a csv of real data,you can get the correlation / covariance matrix via:
    
```python
import pandas as pd

df = pd.read_csv('your_csv_file_here.csv')

# covariance
df.cov()

# correlation matrix
df.corr()
```

- **theory / application problem**
    - there's a few interesting ideas you could ask a question about in some application
        - law of large numbers
        - central limit theorem
        
Choose one of the three bolded problem topics above and write a homework problem below:
- give a clean description which is easily understood by all students
- scaffold your questions from easy to hard
    - if a question, as posed, is too difficult you might consider leading student by breaking it into two
- write a solution for each part, which in addition to being correct, is clear enough to allow students to learn why its correct while reading it

### Part 1 (rubric):
- 3 problem is original / interesting
- 3 problem is appropriately expansive / difficult enough for CS2810 students
    - you can pose something as "extra credit" if its a stretch, but be sure its accesible extra credit
- 3 problem is written clearly and succinctly
- 3 solution is correct
- 3 solution is clear

In [ ]:


import pandas as pd
df = pd.read_csv('datamodels.csv')

# covariance

df.cov()
# correlation matrix


2000        2001        2004       2005        2006      2007
2000  110779.245  115627.455  126406.485  143987.13  151683.075  162391.5
2001  115627.455  120687.845  131938.615  150288.67  158321.425  169498.5
2004  126406.485  131938.615  144238.205  164298.89  173080.475  185299.5
2005  143987.130  150288.670  164298.890  187149.62  197152.550  211071.0
2006  151683.075  158321.425  173080.475  197152.55  207690.125  222352.5
2007  162391.500  169498.500  185299.500  211071.00  222352.500  238050.0

The table below contains two variables, X and Y

X = per capita comsumption of mozzerela cheese.                  
Y = civil engineering doctorates awarded per year

And each column represents a year these variables were measured.


| YEAR | 2000 | 2001 | 2004 | 2005 | 2006 | 2007 |
|------|------|------|------|------|------|------|
| X    | 9.3  | 9.7  | 9.9  | 10.2 | 10.5 | 11   |
| Y    | 480  | 501  | 547  | 622  | 655  | 701  |

Problem 1 : Compute the sample covarience of X and Y




Solution:


In [ ]:
df = df.transpose()

In [ ]:
df.iloc[1:, :].astype(float).cov()

0            1
0   0.364    52.260000
1  52.260  7849.466667

Answer is : 52.26 

Problem 2: Calculate the sample correlation of X and Y. Does this number make sense?

Solution: 
Yes, it makes sense based on the data that the correlation is greater than 0 and close to 1. As X increases, Y also increases.

In [ ]:
df.iloc[1:, :].astype(float).corr()

0         1
0  1.000000  0.977684
1  0.977684  1.000000

Answer is: .978 

Problem 3: Does the increase of mozzerella cheese consumption cause more people to obtain civil engineering doctorates?

Solution:
No, not necesarily. Correlation does not imply causation. Just because X and Y are correlated, doesn't mean one causes the other. 

# Part 2: Build a Bayes Network

In this part, build a homework problem on a Bayes Net on a topic of your choosing.

- Bayes network on four nodes
    - too many nodes and the joint distribution table has too many rows ...
- ask the student 3 or 4 questions which build towards some intuition
    - demonstrate / explain a conditional independence relationship
    - demonstrate some intuition encapsulated in the network:
        - e.g. in HW9 problem 4: "defending a basketball shot decreases the expected points earned"


Your submission should include 4 items:
- the text of the problem
- the solution to each part
- the image of the bayes net (appendix 1)
- the joint distribution table (appendix 2)

# Part 2 (rubric):
- 3 problem is original / interesting
- 3 problem is appropriately expansive / difficult enough for CS2810 students
    - you can pose something as "extra credit" if its a stretch, but be sure its accesible extra credit
- 3 problem is written clearly and succinctly
- 3 solution is correct
- 3 solution is clear
- 3 image of bayes net neat, fully specified and included in jupyter
- 3 markdown table of joint distribution included in jupyter and correct

### Appendix 1: Uploading an image into a jupyter notebook
To create the image of your bayes net 
- be sure to include all probabilities needed to fully specify it
- generate an image:
    - draw one and take a picture with your phone
        - cam scanner please
    - build it on the computer
    - draw on a tablet
- upload into the notebook by:
    - uploading your image to [https://imgbb.com/](https://imgbb.com/)
    - double clicking this jupyter cell
    - replacing the hyperlink below with a hyperlink to your image
        - (should have similar format as the example given)
        - try right clicking "copy image address" or similar

<a href="https://ibb.co/z8jfLVc"><img src="https://i.ibb.co/hXh8nF5/Mini-Proj-Drawing.jpg" alt="Mini-Proj-Drawing" border="0"></a> 

| C: Crying | P(C) | O: Cutting Onions | P(O) | B: Bad Day | P(B) | A: Angry | P(A) | S: Sad | P(S) | P(S,A) | P(S, A, B) | P(S, A, B, O) | P(S,A,B,O,C) |
|-----------|------|-------------------|------|------------|------|----------|------|--------|------|--------|------------|---------------|--------------|
| c0        |  0.9 | o0                |  0.8 | b0         |  0.9 | a0       |  0.7 | s0     |  0.6 |   0.42 |      0.378 |        0.3024 |      0.27216 |
| c0        |  0.9 | o0                |  0.8 | b0         |  0.2 | a0       |  0.7 | s1     |  0.4 |    2.8 |      0.056 |        0.0448 |      0.04032 |
| c0        |  0.9 | o0                |  0.8 | b0         |  0.4 | a1       |  0.3 | s0     |  0.6 |   0.18 |      0.072 |        0.0576 |      0.05184 |
| c0        |  0.9 | o0                |  0.8 | b0         |    0 | a1       |  0.3 | s1     |  0.4 |   0.12 |          0 |             0 |            0 |
| c0        |  0.3 | o0                |  0.8 | b1         |  0.1 | a0       |  0.7 | s0     |  0.6 |   0.42 |      0.042 |        0.0336 |      0.01008 |
| c0        |  0.3 | o0                |  0.8 | b1         |  0.8 | a0       |  0.7 | s1     |  0.4 |    2.8 |      0.224 |        0.1792 |      0.05376 |
| c0        |  0.3 | o0                |  0.8 | b1         |  0.6 | a1       |  0.3 | s0     |  0.6 |   0.18 |      0.108 |        0.0864 |      0.02592 |
| c0        |  0.3 | o0                |  0.8 | b1         |    1 | a1       |  0.3 | s1     |  0.4 |   0.12 |       0.12 |         0.096 |       0.0288 |
| c0        |  0.5 | o1                |  0.2 | b0         |  0.9 | a0       |  0.7 | s0     |  0.6 |   0.42 |      0.378 |        0.0756 |       0.0378 |
| c0        |  0.5 | o1                |  0.2 | b0         |  0.2 | a0       |  0.7 | s1     |  0.4 |    2.8 |      0.056 |        0.0112 |       0.0056 |
| c0        |  0.5 | o1                |  0.2 | b0         |  0.4 | a1       |  0.3 | s0     |  0.6 |   0.18 |      0.072 |        0.0144 |       0.0072 |
| c0        |  0.5 | o1                |  0.2 | b0         |    0 | a1       |  0.3 | s1     |  0.4 |   0.12 |          0 |             0 |            0 |
| c0        |  0.1 | o1                |  0.2 | b1         |  0.1 | a0       |  0.7 | s0     |  0.6 |   0.42 |      0.042 |        0.0084 |      0.00084 |
| c0        |  0.1 | o1                |  0.2 | b1         |  0.8 | a0       |  0.7 | s1     |  0.4 |    2.8 |      0.224 |        0.0448 |      0.00448 |
| c0        |  0.1 | o1                |  0.2 | b1         |  0.6 | a1       |  0.3 | s0     |  0.6 |   0.18 |      0.108 |        0.0216 |      0.00216 |
| c1        |  0.1 | o1                |  0.2 | b1         |    1 | a1       |  0.3 | s1     |  0.4 |   0.12 |       0.12 |         0.024 |       0.0024 |
| c1        |  0.1 | o0                |  0.8 | b0         |  0.9 | a0       |  0.7 | s0     |  0.6 |   0.42 |      0.378 |        0.3024 |      0.03024 |
| c1        |  0.1 | o0                |  0.8 | b0         |  0.2 | a0       |  0.7 | s1     |  0.4 |    2.8 |      0.056 |        0.0448 |      0.00448 |
| c1        |  0.1 | o0                |  0.8 | b0         |  0.4 | a1       |  0.3 | s0     |  0.6 |   0.18 |      0.072 |        0.0576 |      0.00576 |
| c1        |  0.1 | o0                |  0.8 | b0         |    0 | a1       |  0.3 | s1     |  0.4 |   0.12 |          0 |             0 |            0 |
| c1        |  0.7 | o0                |  0.8 | b1         |  0.1 | a0       |  0.7 | s0     |  0.6 |   0.42 |      0.042 |        0.0336 |      0.02352 |
| c1        |  0.7 | o0                |  0.8 | b1         |  0.8 | a0       |  0.7 | s1     |  0.4 |    2.8 |      0.224 |        0.1792 |      0.12544 |
| c1        |  0.7 | o0                |  0.8 | b1         |  0.6 | a1       |  0.3 | s0     |  0.6 |   0.18 |      0.108 |        0.0864 |      0.06048 |
| c1        |  0.7 | o0                |  0.8 | b1         |    1 | a1       |  0.3 | s1     |  0.4 |   0.12 |       0.12 |         0.096 |       0.0672 |
| c1        |  0.5 | o1                |  0.2 | b0         |  0.9 | a0       |  0.7 | s0     |  0.6 |   0.42 |      0.378 |        0.0756 |       0.0378 |
| c1        |  0.5 | o1                |  0.2 | b0         |  0.2 | a0       |  0.7 | s1     |  0.4 |    2.8 |      0.056 |        0.0112 |       0.0056 |
| c1        |  0.5 | o1                |  0.2 | b0         |  0.4 | a1       |  0.3 | s0     |  0.6 |   0.18 |      0.072 |        0.0144 |       0.0072 |
| c1        |  0.5 | o1                |  0.2 | b0         |    0 | a1       |  0.3 | s1     |  0.4 |   0.12 |          0 |             0 |            0 |
| c1        |  0.9 | o1                |  0.2 | b1         |  0.1 | a0       |  0.7 | s0     |  0.6 |   0.42 |      0.042 |        0.0084 |      0.00756 |
| c1        |  0.9 | o1                |  0.2 | b1         |  0.8 | a0       |  0.7 | s1     |  0.4 |    2.8 |      0.224 |        0.0448 |      0.04032 |
| c1        |  0.9 | o1                |  0.2 | b1         |  0.6 | a1       |  0.3 | s0     |  0.6 |   0.18 |      0.108 |        0.0216 |      0.01944 |
| c1        |  0.9 | o1                |  0.2 | b1         |    1 | a1       |  0.3 | s1     |  0.4 |   0.12 |       0.12 |         0.024 |       0.0216 |

1. Given that a student is angry, but not sad, what’s the probability that they’re having a bad day?

- P(S1) = 0.4

2. Given that a student is not sad, having a bad day, and cutting onions, what’s the probability that they’re crying?

- P(B1 | S0A1) = 0.6

3. What has more of an impact on a student crying? Cutting onions or having a bad day?

- P(C1 | B1) > P(C1 | O1)
- P(C1 | B1) = 0.8
- P(C1 | O1) = 0.6

4. What’s the probability a student is having a bad day?

- P(B1) = P(B1 | S0A0) * (S0A0) + P(B1 | S0A1) * (S0A1) + P(B1 | S1A0) * (S1A0) + 
P(B1 | S1A1) * (S1A1) 

- P(B1) = 0.1 * (0.6*0.7) + 0.6 * (0.6*0.3) + 0.8 * (0.4*0.7) + 1 * (0.4*0.3) = 0.494

### Appendix 2: Making a table in a jupyter notebook

- make table in your favorite spreadsheet application
- copy / paste into [https://www.tablesgenerator.com/markdown_tables](https://www.tablesgenerator.com/markdown_tables)
- copy / paste from markdown table generator into a markdown cell in jupyter